api/src/main/java/io/opentelemetry/common/Main.java

```diff
@@ -0,0 +1,139 @@
+ * limitations under the License.
+ */
+
+package io.opentelemetry.common;
+
+import io.opentelemetry.common.CleanReadableAttributes.TypedAttributeConsumer;
+import java.util.List;
+
+@SuppressWarnings({
+  "BadImport",
+  "UnusedVariable",
+  "PrivateConstructorForUtilityClass",
+  "MultipleTopLevelClasses",
+  "JavadocMethod"
+})
+public class Main {
+
+  public static void main(String[] args) {
+    CleanAttributes attributes =
+        CleanAttributes.newBuilder()
+            .setString("string", "I'm a String")
+            .setBoolean("boolean", true)
+            .setDouble("double", 33.444d)
+            .setLong("long", 34333L)
+            .setStringArray("stringArray", "one", "two", "three")
+            .setBooleanArray("booleanArray", true, false, true)
+            .setLongArray("longArray", 33L, 55L, 99L)
+            .setDoubleArray("doubleArray", 123.33, 6655.33, 339393.33, 3434.33)
+            .build();
+
+    System.out.println("Processing with casts:");
+    process(attributes);
+    System.out.println();
+    System.out.println("Processing with types:");
+    processTyped(attributes);
+  }
+
+  private static void processTyped(CleanAttributes attributes) {
+    attributes.forEach(
+        new TypedAttributeConsumer() {
+          @Override
+          public void consumeString(String key, String value) {
+            System.out.println(key + " = " + value);
+          }
+
+          @Override
+          public void consumeLong(String key, long value) {
+            System.out.println(key + " = " + value);
+          }
+
+          @Override
+          public void consumeDouble(String key, double value) {
+            System.out.println(key + " = " + value);
+          }
+
+          @Override
+          public void consumeBoolean(String key, boolean value) {
+            System.out.println(key + " = " + value);
+          }
+
+          @Override
+          public void consumeStringArray(String key, List<String> value) {
+            System.out.println(key + " = " + value);
+          }
+
+          @Override
+          public void consumeLongArray(String key, List<Long> value) {
+            System.out.println(key + " = " + value);
+          }
+
+          @Override
+          public void consumeDoubleArray(String key, List<Double> value) {
+            System.out.println(key + " = " + value);
+          }
+
+          @Override
+          public void consumeBooleanArray(String key, List<Boolean> value) {
+            System.out.println(key + " = " + value);
+          }
+        });
+  }
+
+  public static void process(CleanReadableAttributes attributes) {
+    System.out.println("attributes = " + attributes);
+    attributes.forEach(
+        new CleanReadableAttributes.AttributeConsumer() {
+          @Override
+          public void consume(String key, AttributeType type, Object value) {
+            switch (type) {
+              case STRING:
+                String stringValue = type.asString(value);
```

Comment:
Not strong but I might find this a bit more readable

```java
enum AttributeType {
  STRING, BOOLEAN, FOO, CAT;

  @SuppressWarnings("unchecked")
  public <T> T cast(Object dog) {
    return (T) dog;
  }

case STRING:
  String stringValue = STRING.cast(value);  // Having the same STRING seems easy to read
```

Reply:
I still like my original proposal more :) This code and Anuraag's suggestion are equal to me.

Reply:
I don't know how to implement your proposal. Can you add a PR to this one to show me?

Reply:
@anuraaga it's more readable, for sure, but 100% not type-safe. The user can do the wrong thing trivially, which seems to defeat the purpose, to me. I have this locally, though, and will commit it. :)

Reply:
Is it easier to do the wrong thing with that vs `type.asString`? They still need to call the correct method in either case I think, but using the same `STRING` twice is more obvious I think for a reader of the code since there's symmetry there.

Reply:
Well, using the type capture, you could put any old value on the LHS and it would still compile (`BigDecimal`, `SpanContext`, you name it). So, with the `type.as...` methods, at least they only have a limited number of options?

Unfortunately, at this point we just discussing the least bad option, I'll admit. Java's type system just doesn't have the facilities to make a really good API, without going back to `AttributeValue`, as far as I can see. 

I could probably get behind any of these options, TBH. I think the least-bad option is the 8-way interface, because at least in that case they're forced to consider all the possible types. We can also have the single one and require the manual casting along with that, so at least both options are available, depending on the use-case.

Reply:
+1 vote to 8-way interface (and optionally also single casting variant, but with `toBuilder` the latter becomes less important perhaps)